rskj-core/src/main/java/org/ethereum/datasource/RocksDbKeyIterator.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2018 RSK Labs Ltd.
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.datasource;
+
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksIterator;
+
+import java.util.NoSuchElementException;
+
+public class RocksDbKeyIterator implements DataSourceKeyIterator {
+    private final RocksIterator iterator;
+
+    public RocksDbKeyIterator(RocksDB db) {
+        this(db, false);
+    }
+
+    public RocksDbKeyIterator(RocksDB db, boolean avoidSeekFirst) {
+        this.iterator = db.newIterator();
+
+        if (!avoidSeekFirst) {
+            this.iterator.seekToFirst();
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        this.iterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return this.iterator.isValid();
+    }
+
+    @Override
+    public byte[] next() throws NoSuchElementException {
+        if (!this.hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        this.iterator.next();
```

Comment:
This implementation won't work. Here the first element is always being miss and the last element is never being validated.
An example with the current implementation:

Having an array `[a,b]` during the first iteration when we call  the next() method: 
```
//In the first iteration after call seekToFirst() the iterator is pointing to 'a'
public byte[] next() throws NoSuchElementException {
        if (!this.hasNext()) {  //RocksIterator.isValid() is called returning true
            throw new NoSuchElementException();
        }

        this.iterator.next(); // after this call the iterator is pointing to the next element -> 'b'
        return this.iterator.key(); // 'b' will be returned without being validated. 
    }
```

RocksIterator does not work as the standard Java Iterator. With the java Iterator we call `next()` to get the value of the first element (and move the pointer to the next element) but the next() implementation of the RocksIterator just moves the pointer to the next element and returns void, so we need to get the value before do the next() call. 

The solution would be something like this:
```
public byte[] next() throws NoSuchElementException {
        if (!this.hasNext()) { //validate the current position
            throw new NoSuchElementException(); 
        }
        byte[] currentKey = this.iterator.key();//save the current value 
        this.iterator.next();//advance to the next position
        return currentKey; //return current value
    }
```
In the second call, the RocksIterator is already pointing to the second position which is being validated and returned. In case the array has only one element, after call next() one time the pointer is already pointing to the next element and when we call hasNext() the isValid method will return false, so everything should be ok. 

Sorry for the very long comment.

Reply:
Thanks!

Fixed.